### ISSUE TITLE
Inference with merged LoRA weights

### ISSUE THREAD

**Issue by michele-milesi** [Lightning-AI/lit-gpt#935]
Hi, a few weeks ago @morettif and I finetuned the `Llama70B` with QLoRA on a H100:

- `r=32`
- `alpha=64`
- `quantize=bnb.nf4-dq`
- `precision=bf16-true`
- `weight_decay=0`
- `batch_size=32`
- `micro_batch_size=2`
- `lora_dropout=0.05`
- All LoRA layers

The validation loss after the finetuning was about `0.5`, but during inference on the validation set, we obtained very strange results. We checked the loss on the validation set and it was ~2.5 (as before the finetuning).

We used the `generate/lora.py` script and we noticed that after commenting this line: https://github.com/Lightning-AI/lit-gpt/blob/8a101b633dfeafd378f8fbaba6a80a4417c33576/generate/lora.py#L120 the model generated meaningful output with a loss coherent with the one observed during training (~0.5).

We analyzed the dtype of the pre-trained model weights at the moment of the call to the `merge_lora_weights()` function, some of them are `torch.uint8`, this causes that during merging the following if (in the `merge()` function) is evaluated as `True`: https://github.com/Lightning-AI/lit-gpt/blob/8a101b633dfeafd378f8fbaba6a80a4417c33576/lit_gpt/lora.py#L151

Could there be an error in this piece of code related to the management of quantization?

**Comment by Andrei-Aksionov** [comment]
Got it.
I'll check later this week.

**Comment by soumilchugh** [comment]
Any update on this? I’m encountering similar issues when fine-tuning Llama models (Llama 2 7B and Llama 3.1 8B) with LoRA and 4-bit quantization. During training and validation, everything appears to work smoothly, with training/validation loss dropping as low as 0.5, and the `generate_example` function in `finetune/lora.py` producing consistent outputs. However, during inference with `generate/base.py`, the model gives unexpected results for the same prompt. I’ve tried removing the `merge_lora_weights` function when merging LoRA weights with the pretrained checkpoint, but that hasn’t resolved the issue. Interestingly, LoRA without quantization works perfectly during inference.
